src/relay/analysis/get_calibration_data.cc

```diff
@@ -0,0 +1,187 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/relay/analysis/get_calibration_data.cc
+ *
+ * \brief To get the calibration data, we need to perform two
+ * steps. First, we need to prepare the module that generate
+ * the tensor values (GetCalibrateModule). Second, we need to
+ * generate the mapping between the values and the functions
+ * (GetCalibrateOutputMap).
+ */
+
+#include <tvm/relay/analysis.h>
+#include <tvm/relay/expr.h>
+#include <tvm/relay/expr_functor.h>
+
+namespace tvm {
+namespace relay {
+
+/*!
+ * \brief This function returns a module that will be used by
+ * the relay graph runtime for collecting the calibration data.
+ * To do that, we first make all inputs and outputs of each
+ * function into the final output (i.e., the final output is a
+ * tuple of tensors). Then, we change the compiler attribute of
+ * each function. Finally, we mark all function to be inlined.
+ */
+
+IRModule GetCalibrateModule(IRModule module) {
+  class Collector : public ExprRewriter {
+   public:
+    explicit Collector(const Map<GlobalVar, BaseFunc>& glob_funcs) : glob_funcs_(glob_funcs) {}
+
+    Expr Rewrite_(const CallNode* call, const Expr& post) final {
+      // check if the function implementation is available
+      // intrinsic functions are excluded for now
+      if (call->op->IsInstance<GlobalVarNode>()) {
+        auto var = Downcast<GlobalVar>(call->op);
+        CHECK_GT(glob_funcs_.count(var), 0) << "Function " << var << " is not defined";
+        for (size_t i = 0; i < call->args.size(); i++) new_outputs_.push_back(call->args[i]);
+        // need to flatten the output if it is a tuple
+        auto* fn = glob_funcs_[var].as<FunctionNode>();
+        if (auto* tn = fn->body.as<TupleNode>()) {
+          for (size_t i = 0; i < tn->fields.size(); i++) {
+            new_outputs_.push_back(TupleGetItem(post, i));
+          }
+        } else {
+          new_outputs_.push_back(post);
+        }
+      }
+      return post;
+    }
+
+    Array<Expr> GetNewOutputs() { return new_outputs_; }
+
+   private:
+    const Map<GlobalVar, BaseFunc>& glob_funcs_;
+    Array<Expr> new_outputs_;
+  };
+
+  auto glob_funcs = module->functions;
+  // module is mutable, hence, we make a copy of it.
+  module.CopyOnWrite();
+  for (const auto& pair : glob_funcs) {
+    if (auto* fn = pair.second.as<FunctionNode>()) {
+      auto func = GetRef<Function>(fn);
+      auto* gl_var = pair.first.as<GlobalVarNode>();
+      // we only collect the outputs for main function
+      if (gl_var->name_hint == "main") {
+        Collector collector(glob_funcs);
+        PostOrderRewrite(func->body, &collector);
+        auto new_outputs = collector.GetNewOutputs();
+        if (!new_outputs.empty()) {
+          Array<Expr> fields;
+          for (const auto& output : new_outputs) {
+            fields.push_back(output);
+          }
+          auto tuple = Tuple(fields);
```

Comment:
I think we can do have a `tuple = FlattenOutput(fileds);` here to flatten the output